### PR TITLE
Separate out building and porting docs.

### PR DESCRIPTION
**Pull request by davidchisnall**
Fix omissions in the porting doc (`page_size` and C++20 concepts).

### CODE REVIEW

**Review by mjp41 — APPROVED** [review]

**Comment by davidchisnall** [comment]
@nwf, can you check that the things I've written about `page_size` and the concepts are not nonsense please?

File: docs/BUILDING.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+snmalloc has platform abstraction layers for XNU (macOS, iOS, and so on),
+FreeBSD, NetBSD, OpenBSD, and Linux and is expected to work out of the box on
+these systems.
+Please open issues if it does not.
+Note that NetBSD, by default, ships with a toolchain that emits calls to
+`libatomic` but does not ship `libatomic`.
+To use snmalloc on NetBSD, you must either acquire a `libatomic` implementation
+(for example, from the GCC or LLVM project) or compile with clang.
+
+snmalloc has very few dependencies, CMake, Ninja, Clang 6.0 or later and a C++17
```
Comment (nwf) [inline docs/BUILDING.md:38]:
```suggestion
snmalloc has very few dependencies: CMake, Ninja, Clang 6.0 or later and a C++17
```

File: docs/PORTING.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+```c++
+template<bool committed>
+static void* reserve_aligned(size_t size) noexcept;
+static std::pair<void*, size_t> reserve_at_least(size_t size) noexcept;
+```
+Only one of these needs to be implemented, depending on whether the underlying
+system can provide strongly aligned memory regions.
+If the system guarantees only page alignment, implement the second. The Pal is 
+free to overallocate based on the platforms desire and snmalloc
```
Comment (nwf) [inline docs/PORTING.md:55]:
```suggestion
free to overallocate based on the platform's desire and snmalloc
```

File: docs/BUILDING.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+libatomic for 128-bit atomic operations.
+
+To build a debug configuration:
+```
+mkdir build
+cd build
+cmake -G Ninja .. -DCMAKE_BUILD_TYPE=Debug
+ninja
+```
+To build a release configuration:
```
Comment (nwf) [inline docs/BUILDING.md:50]:
```suggestion
To build a release configuration (that is, one with optimizations turned on):
```

File: docs/BUILDING.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+`LD_PRELOAD`ed to use the allocator in place of the system allocator, for
+example, you can run the build script using the snmalloc as the allocator for
+your toolchain:
+
+```
+LD_PRELOAD=/usr/local/lib/libsnmallocshim.so ninja
+```
+
+## Cross Compile for Android
+Android support is out-of-the-box.
```
Comment (nwf) [inline docs/BUILDING.md:76]: Maybe?

```suggestion
Android is supported out-of-the-box.
```

File: docs/PORTING.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+will find suitably aligned blocks inside the region.  `reserve_at_least` should 
+not commit memory as snmalloc will commit the range of memory it requires of what 
+is returned.
+
+If the system provides strong alignment, implement the first to return memory
+at the desired alignment. If providing the first, then the `Pal` should also 
+specify the minimum size block it can provide: 
+```
+static constexpr size_t minimum_alloc_size = ...;
+```
```
Comment (nwf) [inline docs/PORTING.md:65]: This is not at present enforced by the concept, but probably should be.

File: docs/PORTING.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+The PAL is also responsible for advertising the page size:
+
+```c++
+static constexpr size_t page_size = 0x1000;
+```
+
+This is the size at which the PAL is able to mark memory as in-use or not-in-use.
```
Comment (nwf) [inline docs/PORTING.md:73]:
```suggestion
This is the granularity at which the PAL is able to mark memory as in-use or not-in-use.
```

**Review by nwf — COMMENTED** [review]
Some nits, but otherwise LGTM.

File: docs/BUILDING.md (hunk 3 above)
Comment (davidchisnall) [reply docs/BUILDING.md:50]: Release also implies no asserts and no debug info, so I think this is a misleading change.

Comment (nwf) [reply docs/BUILDING.md:50]: Are asserts implied by `RelWithDebugInfo`?  Clarifying both branches might be worthwhile.

Comment (davidchisnall) [reply docs/BUILDING.md:50]: No, asserts are based on the NDEBUG macro not being defined.  All of this is inherited from CMake and I'm hesitant to try to redefine core CMake configs in our build.

Comment (nwf) [reply docs/BUILDING.md:50]: Sorry, to be clear: I wasn't suggesting changing anything about CMake or snmalloc's use of it, I was just reacting to the next stanza in the docs, which describe `RelWithDebugInfo` as "with optimizations on but with debug info".  That doesn't make it clear whether it's Release-like in having shed assertions or Debug-like in having kept them.

**Pull request merged**